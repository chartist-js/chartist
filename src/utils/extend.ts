/**
 * Simple recursive object extend
 * @param target Target object where the source will be merged into
 * @param sources This object (objects) will be merged into target and then target is returned
 * @return An object that has the same reference as target but is extended and merged with the properties of source
 */
export function extend<T>(target: T): T;
export function extend<T, A>(target: T, a: A): T & A;
export function extend<T, A, B>(target: T, a: A, b: B): T & A & B;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function extend(target: any = {}, ...sources: any[]) {
  for (let i = 0; i < sources.length; i++) {
    const source = sources[i];
    const targetProto = Object.getPrototypeOf(target);
    for (const prop in source) {
      if (targetProto !== null && prop in targetProto) {
        continue; // prevent prototype pollution
      }
      const sourceProp = source[prop];
      if (
        typeof sourceProp === 'object' &&
        sourceProp !== null &&
        !(sourceProp instanceof Array)
      ) {
        target[prop] = extend(target[prop], sourceProp);
      } else {
        target[prop] = sourceProp;
      }
    }
  }

  return target;
}
