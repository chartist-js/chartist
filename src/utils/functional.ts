/**
 * Helps to simplify functional style code
 * @param n This exact value will be returned by the noop function
 * @return The same value that was provided to the n parameter
 */
export const noop = <T>(n: T) => n;

/**
 * Functional style helper to produce array with given length initialized with undefined values
 */
export function times(length: number): undefined[];
export function times<T = unknown>(
  length: number,
  filler: (index: number) => T
): T[];
export function times<T = unknown>(
  length: number,
  filler?: (index: number) => T
) {
  return Array.from({ length }, filler ? (_, i) => filler(i) : () => void 0);
}

/**
 * Sum helper to be used in reduce functions
 */
export const sum = (previous: number, current: number) =>
  previous + (current ? current : 0);

/**
 * Map for multi dimensional arrays where their nested arrays will be mapped in serial. The output array will have the length of the largest nested array. The callback function is called with variable arguments where each argument is the nested array value (or undefined if there are no more values).
 *
 * For example:
 * @example
 * ```ts
 * const data = [[1, 2], [3], []];
 * serialMap(data, cb);
 *
 * // where cb will be called 2 times
 * // 1. call arguments: (1, 3, undefined)
 * // 2. call arguments: (2, undefined, undefined)
 * ```
 */
export const serialMap = <T, K>(array: T[][], callback: (...args: T[]) => K) =>
  times(Math.max(...array.map(element => element.length)), index =>
    callback(...array.map(element => element[index]))
  );
