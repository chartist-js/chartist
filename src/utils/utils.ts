import type { FilterByKey } from './types';

/**
 * This function safely checks if an objects has an owned property.
 * @param target The object where to check for a property
 * @param property The property name
 * @returns Returns true if the object owns the specified property
 */
export function safeHasProperty<T, K extends string>(
  target: T,
  property: K
): target is FilterByKey<T, K>;
export function safeHasProperty(target: unknown, property: string) {
  return (
    target !== null &&
    typeof target === 'object' &&
    Reflect.has(target, property)
  );
}

/**
 * Checks if a value can be safely coerced to a number. This includes all values except null which result in finite numbers when coerced. This excludes NaN, since it's not finite.
 */
export function isNumeric(value: number): true;
export function isNumeric(value: unknown): boolean;
export function isNumeric(value: unknown) {
  return value !== null && isFinite(value as number);
}

/**
 * Returns true on all falsey values except the numeric value 0.
 */
export function isFalseyButZero(
  value: unknown
): value is undefined | null | false | '' {
  return !value && value !== 0;
}

/**
 * Returns a number if the passed parameter is a valid number or the function will return undefined. On all other values than a valid number, this function will return undefined.
 */
export function getNumberOrUndefined(value: number): number;
export function getNumberOrUndefined(value: unknown): number | undefined;
export function getNumberOrUndefined(value: unknown) {
  return isNumeric(value) ? Number(value) : undefined;
}

/**
 * Checks if value is array of arrays or not.
 */
export function isArrayOfArrays(data: unknown): data is unknown[][] {
  if (!Array.isArray(data)) {
    return false;
  }

  return data.every(Array.isArray);
}

/**
 * Loop over array.
 */
export function each<T>(
  list: T[],
  callback: (item: T, index: number, itemIndex: number) => void,
  reverse = false
) {
  let index = 0;

  list[reverse ? 'reduceRight' : 'reduce']<void>(
    (_, item, itemIndex) => callback(item, index++, itemIndex),
    void 0
  );
}
