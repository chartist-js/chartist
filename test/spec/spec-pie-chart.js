import 'utils/dom-prepare-before-each';
import 'utils/dom-cleanup-after-each';

import Chartist from '../../dist/chartist';

import { appendToDom, cleanupDom } from 'utils/dom';

describe('Pie chart tests', function() {
  'use strict';

  beforeEach(function() {

  });

  afterEach(function() {

  });

  describe('Meta data tests', function() {
    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');
    });

    afterEach(function () {
      cleanupDom();
    });

    it('should render meta data correctly on slice with mixed value array', function (done) {
      var meta = {
        test: 'Serialized Test'
      };

      var data = {
        labels: ['A', 'B', 'C'],
        series: [5, {
          value: 8,
          meta: meta
        }, 1]
      };

      var chart = new Chartist.Pie('.ct-chart', data);

      chart.on('created', function() {
        const slices = document.querySelectorAll('.ct-slice-pie');

        expect(Chartist.deserialize(slices[1].getAttribute('ct:meta'))).toEqual(meta);
        done();
      });
    });
  });

  describe('Simple Pie Chart', function() {
    // https://gionkunz.github.io/chartist-js/examples.html#simple-pie-chart

    var num = '\\d+(\\.\\d*)?';
    var data, options;

    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');

      var sum = function(a, b) { return a + b; };
      data = {
        series: [5, 3, 4]
      };
      options = {
        width: 100,
        height: 100,
        chartPadding: 10,
        labelInterpolationFnc: function(value) {
          return Math.round(value / data.series.reduce(sum) * 100) + '%';
        }
      };
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var chart = new Chartist.Pie('.ct-chart', data, options);
      chart.on('created', callback);
    }

    it('should render three slices', function(done) {
      onCreated(function() {
        const slices = document.querySelectorAll('.ct-slice-pie');

        expect(slices.length).toBe(3);
        done();
      });
    });

    it('should set value attribute', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');

        expect(slices[0].getAttribute('ct:value')).toBe('5');
        expect(slices[1].getAttribute('ct:value')).toBe('3');
        expect(slices[2].getAttribute('ct:value')).toBe('4');
        done();
      });
    });

    it('should create slice path', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');

        [].slice.call(slices).forEach(function(el) {
          var pattern =
            '^M' + num + ',' + num +
            'A40,40,0,0,0,' + num + ',' + num +
            'L50,50Z$';
          var path = el.getAttribute('d');
          expect(path).toMatch(new RegExp(pattern));
        });
        done();
      });
    });

    it('should add labels', function(done) {
      onCreated(function() {
        var labels = document.querySelectorAll('.ct-label');

        expect(labels[0].textContent).toContain('42%');
        expect(labels[1].textContent).toContain('25%');
        expect(labels[2].textContent).toContain('33%');
        done();
      });
    });

    it('should overlap slices', function(done) {
      data = {
        series: [1, 1]
      };
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[0];
        var slice2 = slices[1];

        expect(slice1.getAttribute('d')).toMatch(/^M50,90A40,40,0,0,0,50,10L50,50Z/);
        expect(slice2.getAttribute('d')).toMatch(/^M50,10A40,40,0,0,0,50.\d+,90L50,50Z/);
        done();
      });
    });

    it('should set large arc sweep flag', function(done) {
      data = {
        series: [1, 2]
      };
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[1];
        
        expect(slice1.getAttribute('d')).toMatch(/^M50,10A40,40,0,1,0/);
        done();
      }, data);
    });

    it('should draw complete circle with gap', function(done) {
      data = {
        series: [1]
      };
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[0];
        expect(slice1.getAttribute('d')).toMatch(/^M49.9\d+,10A40,40,0,1,0,50,10L50,50Z/);
        done();
      });
    });

    it('should draw complete circle with startAngle', function(done) {
      data.series = [100];
      options.startAngle = 90;
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[0];
        expect(slice1.getAttribute('d')).toMatch(/^M90,49.9\d+A40,40,0,1,0,90,50L50,50Z/);
        done();
      });
    });

    it('should draw complete circle if values are 0', function(done) {
      data = {
        series: [0, 1, 0]
      };
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[1];
        expect(slice1.getAttribute('d')).toMatch(/^M49.9\d+,10A40,40,0,1,0,50,10L50,50Z/);
        done();
      });
    });

  });

  describe('Pie with small slices', function() {
    var data, options;

    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');

      data = {
        series: [0.001, 2]
      };
      options =  {
        width: 100,
        height: 100,
        chartPadding: 0
      };
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var chart = new Chartist.Pie('.ct-chart', data, options);
      chart.on('created', callback);
    }

    it('Pie should render correctly with very small slices', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[0];
        var slice2 = slices[1];

        expect(slice1.getAttribute('d')).toMatch(/^M50.1\d+,0A50,50,0,0,0,50,0/);
        expect(slice2.getAttribute('d')).toMatch(/^M49.9\d*,0A50,50,0,1,0,50,0/);
        done();
      });
    });

    it('Pie should render correctly with very small slices on startAngle', function(done) {
      options.startAngle = 90;
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');
        var slice1 = slices[0];
        var slice2 = slices[1];

        expect(slice1.getAttribute('d')).toMatch(/^M100,50.1\d*A50,50,0,0,0,100,50/);
        expect(slice2.getAttribute('d')).toMatch(/^M100,49.97\d*A50,50,0,1,0,100,49.98\d*/);
        done();
      });
    });

    it('Donut should render correctly with very small slices', function(done) {
      options.donut = true;
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-donut');
        var slice1 = slices[0];
        var slice2 = slices[1];

        expect(slice1.getAttribute('d')).toMatch(/^M50.\d+,30A20,20,0,0,0,50,30/);
        expect(slice2.getAttribute('d')).toMatch(/^M49.9\d*,30A20,20,0,1,0,50,30/);
        done();
      });
    });

  });

  describe('Pie with some empty values configured to be ignored', function() {
    var data, options;

    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');

      data = {
        series: [1, 2, 0, 4]
      };
      options =  {
        width: 100,
        height: 100,
        ignoreEmptyValues: true
      };
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var chart = new Chartist.Pie('.ct-chart', data, options);
      chart.on('created', callback);
    }

    it('Pie should not render empty slices', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');

        expect(slices.length).toBe(3);

        expect(slices[0].getAttribute('ct:value')).toBe('1');
        expect(slices[1].getAttribute('ct:value')).toBe('2');
        expect(slices[2].getAttribute('ct:value')).toBe('4');
        done();
      });
    });
  });

  describe('Pie with empty values', function() {
    var data;

    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');

      data = {
        series: [0, 0, 0]
      };
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var chart = new Chartist.Pie('.ct-chart', data, {});
      chart.on('created', callback);
    }

    it('Pie should render without NaN values and points', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');

        expect(slices.length).toBe(3);

        expect(slices[0].getAttribute('ct:value')).toBe('0');
        expect(slices[1].getAttribute('ct:value')).toBe('0');
        expect(slices[2].getAttribute('ct:value')).toBe('0');

        expect(slices[0].getAttribute('d')).toMatchSnapshot();
        expect(slices[1].getAttribute('d')).toMatchSnapshot();
        expect(slices[2].getAttribute('d')).toMatchSnapshot();
        done();
      });
    });
  });

  describe('Pie with some empty values configured not to be ignored', function() {
    var data, options;

    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');

      data = {
        series: [1, 2, 0, 4]
      };
      options =  {
        width: 100,
        height: 100,
        ignoreEmptyValues: false
      };
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var chart = new Chartist.Pie('.ct-chart', data, options);
      chart.on('created', callback);
    }

    it('Pie should render empty slices', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-pie');

        expect(slices.length).toBe(4);

        expect(slices[0].getAttribute('ct:value')).toBe('1');
        expect(slices[1].getAttribute('ct:value')).toBe('2');
        expect(slices[2].getAttribute('ct:value')).toBe('0');
        expect(slices[3].getAttribute('ct:value')).toBe('4');
        done();
      });
    });
  });

  describe('Gauge Chart', function() {
    // https://gionkunz.github.io/chartist-js/examples.html#gauge-chart
    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var data = {
        series: [20, 10, 30, 40]
      };
      var options =  {
        chartPadding:50,
        height:500,
        width:500,
        donut: true,
        donutWidth: 60,
        startAngle: 270,
        total: 200,
        showLabel: false
      };
      var chart = new Chartist.Pie('.ct-chart', data, options);
      chart.on('created', callback);
    }

    it('should render four strokes', function(done) {
      onCreated(function() {
        expect(document.querySelectorAll('.ct-slice-donut').length).toBe(4);
        done();
      });
    });

    it('should set value attribute', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-donut');

        expect(slices[0].getAttribute('ct:value')).toBe('20');
        expect(slices[1].getAttribute('ct:value')).toBe('10');
        expect(slices[2].getAttribute('ct:value')).toBe('30');
        expect(slices[3].getAttribute('ct:value')).toBe('40');
        done();
      });
    });

    it('should create slice path', function(done) {
      onCreated(function() {
        var slices = document.querySelectorAll('.ct-slice-donut');

        [].slice.call(slices).forEach(function(el) {
          var num = '\\d+(\\.\\d*)?';
          var pattern =
            '^M' + num + ',' + num +
            'A170,170,0,0,0,' + num + ',' + num +
            '$';
          var path = el.getAttribute('d');
          expect(path).toMatch(new RegExp(pattern));
        });
        done();
      });
    });

    it('should set stroke-width', function(done) {
     onCreated(function() {
      var slices = document.querySelectorAll('.ct-slice-donut');

      [].slice.call(slices).forEach(function(el) {
          var style = el.getAttribute('style');
          expect(style).toMatch(new RegExp('stroke-width:\\s?60px'));
        });
        done();
      });
    });

    it('should not add labels', function(done) {
      onCreated(function() {
        var labels = document.querySelectorAll('.ct-label');
        expect(labels.length).toBe(0);
        done();
      });

    });

  });

  describe('Pie Chart with relative donutWidth', function() {
    beforeEach(function() {
      appendToDom('<div class="ct-chart ct-golden-section"></div>');
    });

    afterEach(function () {
      cleanupDom();
    });

    function onCreated(callback) {
      var data = {
        series: [20, 10, 30, 40]
      };
      var options =  {
        chartPadding:50,
        height:500,
        width:500,
        donut: true,
        donutWidth: '25%',
        showLabel: false
      };
      var chart = new Chartist.Pie('.ct-chart', data, options);
      chart.on('created', callback);
    }

    it('should render four strokes', function(done) {
      onCreated(function() {
        expect(document.querySelectorAll('.ct-slice-donut').length).toBe(4);
        done();
      });
    });

    it('should create slice path', function(done) {
      onCreated(function() {
        [].slice.call(document.querySelectorAll('.ct-slice-donut')).forEach(function(el) {

          var num = '\\d+(\\.\\d*)?';
          var pattern =
            '^M' + num + ',' + num +
            'A175,175,0,0,0,' + num + ',' + num +
            '$';
          var path = el.getAttribute('d');
          expect(path).toMatch(new RegExp(pattern));
        });
        done();
      });
    });

    it('should set stroke-width', function(done) {
     onCreated(function() {
        [].slice.call(document.querySelectorAll('.ct-slice-donut')).forEach(function(el) {
          var style = el.getAttribute('style');
          expect(style).toMatch(new RegExp('stroke-width:\\s?50px'));
        });
        done();
      });
    });
  });

});
