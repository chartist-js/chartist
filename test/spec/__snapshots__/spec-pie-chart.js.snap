// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pie chart tests Pie with empty values Pie should render without NaN values and points 1`] = `"M5.5,5A0.5,0.5,0,0,0,5.5,5L5.5,5.5Z"`;

exports[`Pie chart tests Pie with empty values Pie should render without NaN values and points 2`] = `"M5.5,5A0.5,0.5,0,0,0,5.5,5L5.5,5.5Z"`;

exports[`Pie chart tests Pie with empty values Pie should render without NaN values and points 3`] = `"M5.5,5A0.5,0.5,0,0,0,5.5,5L5.5,5.5Z"`;
